ggu/main.c

```diff
@@ -100,23 +108,130 @@ process_single_policy(const struct ggu_policy *policy, const struct ggu_config *
+					block, __func__);
+				return;
+			}
+			policy.state = GK_DECLINED;
+			policy.flow.proto = ETHER_TYPE_IPv4;
+			rte_memcpy(&policy.flow.f.v4, ggu_decision->ip_flow,
+				sizeof(policy.flow.f.v4));
+			rte_memcpy(&policy.params.u.declined,
+				ggu_decision->ip_flow +
+				sizeof(policy.flow.f.v4),
+				sizeof(policy.params.u.declined));
+			break;
+		case GGU_DEC_IPV6_DECLINED:
+			decision_len =  sizeof(*ggu_decision) +
+				sizeof(policy.flow.f.v6) +
+				sizeof(policy.params.u.declined);
+			if (decision_list_len < decision_len) {
+				RTE_LOG(WARNING, GATEKEEPER,
+					"%s: %s: malformed IPv6 declined decision\n",
+					block, __func__);
+				return;
+			}
+			policy.state = GK_DECLINED;
+			policy.flow.proto = ETHER_TYPE_IPv6;
+			rte_memcpy(&policy.flow.f.v6, ggu_decision->ip_flow,
+				sizeof(policy.flow.f.v6));
+			rte_memcpy(&policy.params.u.declined,
+				ggu_decision->ip_flow +
+				sizeof(policy.flow.f.v6),
+				sizeof(policy.params.u.declined));
+			break;
+		case GGU_DEC_IPV4_GRANTED:
+			decision_len =  sizeof(*ggu_decision) +
+				sizeof(policy.flow.f.v4) +
+				sizeof(policy.params.u.granted);
+			if (decision_list_len < decision_len) {
+				RTE_LOG(WARNING, GATEKEEPER,
+					"%s: %s: malformed IPv4 granted decision\n",
+					block, __func__);
+				return;
+			}
+			policy.state = GK_GRANTED;
+			policy.flow.proto = ETHER_TYPE_IPv4;
+			rte_memcpy(&policy.flow.f.v4, ggu_decision->ip_flow,
+				sizeof(policy.flow.f.v4));
+			rte_memcpy(&policy.params.u.granted,
+				ggu_decision->ip_flow +
+				sizeof(policy.flow.f.v4),
+				sizeof(policy.params.u.granted));
+			break;
+		case GGU_DEC_IPV6_GRANTED:
+			decision_len =  sizeof(*ggu_decision) +
+				sizeof(policy.flow.f.v6) +
+				sizeof(policy.params.u.granted);
+			if (decision_list_len < decision_len) {
+				RTE_LOG(WARNING, GATEKEEPER,
+					"%s: %s: malformed IPv6 granted decision\n",
+					block, __func__);
+				return;
+			}
+			policy.state = GK_GRANTED;
+			policy.flow.proto = ETHER_TYPE_IPv6;
+			rte_memcpy(&policy.flow.f.v6, ggu_decision->ip_flow,
+				sizeof(policy.flow.f.v6));
+			rte_memcpy(&policy.params.u.granted,
+				ggu_decision->ip_flow +
+				sizeof(policy.flow.f.v6),
+				sizeof(policy.params.u.granted));
+			break;
+		default:
+			RTE_LOG(WARNING, GATEKEEPER,
+				"%s: %s: unexpected decision type: %hu\n",
+				block, __func__, decision_type);
+			return;
+		}
+
+		policy_fn(&policy, policy_arg);
+		ggu_decision = (struct ggu_decision *)
+			(((uint8_t *)ggu_decision) + decision_len);
+		decision_list_len -= decision_len;
+	}
+
+	if (decision_list_len != 0) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"%s: %s: notification packet had partial decision list\n",
+			block, __func__);
+	}
+}
+
+static void
+process_single_packet(struct rte_mbuf *pkt, struct ggu_config *ggu_conf)
 {
-	uint8_t j;
-	uint8_t *policy_ptr;
 	uint16_t ether_type;
 	struct ether_hdr *eth_hdr;
 	void *l3_hdr;
 	struct ipv4_hdr *ip4hdr;
 	struct ipv6_hdr *ip6hdr;
```

Comment:
Let's move `ip4hdr`  and `ip6hdr` to their respective `case` in the `switch` below.